### PR TITLE
Collect mocha tests using mocha

### PR DESCRIPTION
**Pull request by mattfysh**

### CODE REVIEW

File: src/interop/lib/get_mocha_tests.js [hunk 1]
```diff
@@ -1,68 +1,20 @@
...
 
-var mochaSettings = require("./mocha_settings");
+var reporter = path.resolve(__dirname, 'test_capture.js');
+var mochaSettings = require('./mocha_settings');
+var mochaOpts = mochaSettings.mochaOpts;
+var mochaAtom = mochaSettings.mochaAtom;
 
-var sourceFolders = mochaSettings.mochaTestFolders;
+module.exports = function() {
+  var cmd = 'mocha --reporter ' + reporter + ' --opts ' + mochaOpts;
```
Comment (Maciek416) [inline src/interop/lib/get_mocha_tests.js]: I think this line might assume mocha is in PATH, we may want to go to `node_modules/.bin`.

File: src/interop/lib/get_mocha_tests.js [hunk 2]
```diff
@@ -1,68 +1,20 @@
...
-var path = require("path");
-var fs = require("fs");
+var path = require('path');
+var exec = require('sync-exec');
 
-var mochaSettings = require("./mocha_settings");
+var reporter = path.resolve(__dirname, 'test_capture.js');
+var mochaSettings = require('./mocha_settings');
+var mochaOpts = mochaSettings.mochaOpts;
+var mochaAtom = mochaSettings.mochaAtom;
```
Comment (Maciek416) [inline src/interop/lib/get_mocha_tests.js]: still TODO I assume?

**Comment by Maciek416** [comment]
This is looking pretty good! I'm excited about enumerating mocha tests in a more reliable way.

File: src/interop/lib/get_mocha_tests.js (hunk 2 above)
Comment (mattfysh) [inline src/interop/lib/get_mocha_tests.js]: that's the next task - inspecting mochaAtom and then using this to segment the test suite. Aiming for 'browser' and 'test' to begin with, other possible future values could be 'file' or 'suite'. Feel free to suggest a new name for the configuration option :)

**Comment by geekdave** [comment]
@mattfysh Thanks so much for the contribution!  Let us know how we can help to get this merge-ready! :zap:

**Comment by mattfysh** [comment]
no problem @geekdave ! I'm a little stuck with handling retries in the event of a failure, as you probably don't want to run the entire suite again if only 2 or 3 specs failed. What I think I'll do is send this pull request through as-is to help improve mocha support, and we can tackle distribution & retries in a follow up - how does that sound?

**Comment by ztsmith** [comment]
Will this PR address the issue of mocha [root-level hooks](https://mochajs.org/#root-level-hooks) not being run by Magellan?
They seem to be ignored with the current version.

**Comment by mattfysh** [comment]
@ztsmith it will, I'm planning to submit final PR this weekend.

**Comment by Maciek416** [comment]
I'd be interested in seeing a simple test suite that has root level hook that doesn't run and have it be part of our test suite from now on.

**Comment by mattfysh** [comment]
@Maciek416 perfect, I'll add that to this branch :)

**Comment by mattfysh** [comment]
@Maciek416 finished functionality, will begin writing tests. added support for describe.only/it.only, as well as now reporting on pending tests (xdescribe/xit) - let me know if it functionally looks good to go, and I'll finish up the PR including tests by tomorrow. cheers :)

**Comment by Maciek416** [comment]
This is looking pretty awesome.

With the introduction of `pending`, we may be looking at a few regressions in other [admittedly undertested] places across magellan codebase. Namely, I'm curious what effect this will have on computing the early bailout (`--bail_early`).

Also, I haven't given any deep thinking to how our `admiral` reporter will handle pending status, though I **believe** this is a less urgent issue because we appear to never run the test in the first place, so the admiral reporter should never become aware of the pending test in the first place:

```
    if (test.path.pending) {
      test.pending();
      onTestComplete(null, test);
```

If we land this, perhaps in the future we'll formulate a new visual state in our `admiral` dashboard to indicate that a test is pending. Could be useful.

**Comment by Maciek416** [comment]
(Just to be clear, please don't consider the above `admiral` notes to be a blocker for this PR. If you do have time, I'd appreciate if you could double check the impact on `--bail_early` computation). :)

**Comment by mattfysh** [comment]
had to remove the --worker and --mocking_port args sent to the workers as commander threw an error if these were present with no args (the test.path.file was removed so the full mocha suite is loaded incl. root level hooks)

**Comment by mattfysh** [comment]
@Maciek416 I believe this is finished, I just need to integrate with our application to see if there are any remaining gaps. feel free to jump in early and provide feedback, thanks :)

**Comment by mattfysh** [comment]
Also I'd checked the code for bail early and the size of the pending array doesn't factor into the ratio calc, so I think we're good there.

**Comment by Maciek416** [comment]
Thanks for checking.

**Comment by mattfysh** [comment]
@Maciek416 any tips to move this PR forward? thanks!

**Comment by Maciek416** [comment]
I **think** the only thing holding us back here is the ES6 fix that @alexlande did, but since your feature removes our AST parsing requirement via Acorn, that becomes a non-issue with your PR.

Can you check what's conflicting and give me a shout ? I'd like to unblock you here.

**Comment by Maciek416** [comment]
In the spirit of reducing the irony factor of this project that you pointed out elsewhere, I've added some tests for the test iteration on the nightwatch side of things. We should probably add similar tests here.

**Comment by kabirsikand** [comment]
+1 This would be very helpful for everyone using coffeescript

**Comment by adstage-jason** [comment]
It looks like this PR solves some of the pain points that we've been running into, so really appreciate the hard work here! Is there any timeline on moving this PR forward?

**Comment by spinlock99** [comment]
:+1: I'd love to see this get merged :)

**Comment by Maciek416** [comment]
Hey all, I've been moving our test framework support to plugins and should be landing that work either this or next week (it's done and working now, I'm just making sure plugins can register with the `--help` output and some documentation around how to actually include these plugins).

I'd like to see @mattfysh 's PR land on the mocha plugin as soon as possible thereafter. Also, this will make it easier to work on test framework support at a quicker pace than `magellan` itself, as well as make it easier to write tests.

If y'all wanna follow along and crack the whip / bang the drum ( :smile: ) be sure to check out https://github.com/TestArmada/magellan/pull/91

**Comment by Maciek416** [comment]
BTW @mattfysh, I'd be happy to jam this over myself if you're too busy, but I'd prefer for you to nab the contributor credits if possible. Let me know if things get too hairy though

**Comment by mattfysh** [comment]
@Maciek416 I've been holding back a bit waiting for the outcome on the interop -> plugin move. should we abandon this PR and move this over to the other mocha plugin repo? also love to see that rowdy and vanilla mocha are now one. as far as this repo goes, the main thing that needs to stay here is probably the work on pending states.

**Comment by Maciek416** [comment]
@mattfysh Yep, I think that's what we should do. The plugin structure isn't going to change much (aside from CLI help being added) so it's wide open.

**Comment by Maciek416** [comment]
Barring any surprises from the day job (it's performance review season!), I am really hoping to land plugin support tomorrow. The `mocha` plugin (here: https://github.com/TestArmada/magellan-mocha-plugin ) is ready for your changes from this PR if you want to post something up.

**Comment by mattfysh** [comment]
ha, we also just had performance reviews! I'm still in the middle of shifting everything to the plugin model, but need to get this done before monday... so guess what I'm doing this weekend ;)

**Comment by Maciek416** [comment]
Alright, `magellan` `5.x.x` has landed. Let's dust this one off!

**Comment by Maciek416** [comment]
Heads up @mattfysh : I am changing how path/filename/etc are set and structured in the plugins and in the `test_runner`. Hopefully the last major structural change to plugins I have to make for a while

**Comment by mattfysh** [comment]
@Maciek416 ok, is there a branch I can look at? Sorry for the delay!

**Comment by Maciek416** [comment]
@mattfysh yeah, here's how the plugin is morphing: https://github.com/TestArmada/magellan-mocha-plugin/pull/2

**Comment by mattfysh** [comment]
Closing this in favor of #100 and https://github.com/TestArmada/magellan-mocha-plugin/pull/3

**Pull request closed without merging**